rfc-154-integrating-an-ots-cms-with-govuk-for-custom-pages.md

```diff
@@ -0,0 +1,60 @@
+
+- no increase in dependency for GOV.UK rendering applications, they continue to receive page content from Content Store and do not need to consider CMS unavailability
+- ability to integrate a page with Search API using existing Publishing API message queue approach
+- implicit updating of Router API
+
+However in order to facilitate this there would need to be new software built, to bridge communication between a CMS and GOV.UK, and changes to existing software to facilitate this bridging.
+
+### Bridging software
+
+It is common that headless CMS tools make use of [webhooks](https://en.wikipedia.org/wiki/Webhook) as a means to notify subscribing software to events that occur in the CMS. We propose the introduction of a piece of software that will listen for a CMS's webhooks and use this information as an initiator of a process to update the content.
+
+This software will identify whether a webhook event signifies a live or draft event and update the Publishing API accordingly.
+
+For integration with Contentful, we will deploy a new GOV.UK application: Contentful Listener API. This was [developed](https://github.com/alphagov/contentful-listener-api) whilst spiking approaches to this problem. There is documentation written to the describe the [workflow](https://github.com/alphagov/contentful-listener-api/blob/main/docs/how-this-application-works.md) of this tool and [integration limitations](https://github.com/alphagov/contentful-listener-api/blob/main/docs/integration-limitations.md). This application will be owned by the Publishing Platform team.
+
+### Changes to existing software
+
+It is expected that in order to flexibly model a GOV.UK page in an off-the-shelf headless CMS there will need to be multiple content entities involved (ie. to model the Cost of living page, there may be multiple accordion entities associated with a Cost of living page entity). In order to track whether these are associated with content a new field will be added to the Publishing API: `cms_entity_ids` this can be used to store the unique IDs of items in the off-the-shelf headless CMS. When a webhook event is received this ID will be queried to determine if any GOV.UK content is affected.
+
+The [`special_route`](https://github.com/alphagov/govuk-content-schemas/blob/main/formats/special_route.jsonnet) schema in [govuk-content-schemas](https://github.com/alphagov/govuk-content-schemas) will be updated to accept a details field. This details field will not have restrictions on how the data is structured, this will allow changes in the modelling of content without needing a schema to be updated (this was an [approach explored](https://github.com/alphagov/govuk-content-schemas/blob/170a941b42bbc53d529b4c50698ad5ba50776df5/formats/coronavirus_landing_page.jsonnet#L4-L8) for the Coronavirus landing page). It is theorised that having a schema for a custom page is more of a hindrance than benefit given the data is only used in a single place.
```

Comment:
Is it right that we modify the existing special route schema, rather than add a new, generic one for externally generated content? I take your point about it only being used in a single place (for now), but I guess there's a reason that existing special routes explicitly cannot have details sections, so modifying them maybe feels a bit wrong.

Reply:
The existing ones don't have a details section as they are typically hardcoded content. 

We’ve got around this in the past with creating new schemas such for handling freeform data like [coronavirus_landing_page](https://github.com/alphagov/govuk-content-schemas/blob/cdc8bfe8b13a24285b7dcae99248adb1dd31de29/formats/coronavirus_landing_page.jsonnet).

I’d argue that these pages however are very much in common with most special routes as they are indeed special, and unlikely other schemas, don’t have a fixed data structure. Most special routes are used for custom GOV.UK pages and these will also be custom GOV.UK pages.

Reply:
I'm not sure if the coronavirus_landing_page is a good example. I think we used [Specialist Route Publisher](https://github.com/alphagov/special-route-publisher/pull/84/files) to publish the landing page, but its schema was never `special_route` but `coronavirus_landing_page`. 

Also there's a niggle in the back of my brain that says special routes by default aren't indexed by search. I haven't been able to find concrete evidence of this though, just [this entry](https://github.com/alphagov/search-api/blob/babb620cd2918ff3fc918fce5c21a07ce9b3ecd8/config/govuk_index/migrated_formats.yaml#L320), which seems to suggest only particular pages are not indexed. But perhaps the indexable-ness of special routes should be confirmed?